File: src/main/resources/js/bst.js
function BST(data) {
    this.root = new Node(data);
}

function Node(data) {
    this.data = data;
    this.left = null;
    this.right = null;
}

Node.prototype.add = function(data) {
    if (data && this.data !== data) {
        if (data < this.data) {
            if (this.left) {
                this.left.add(data);
            } else {
                this.left = new Node(data);
            }
        } else {
            if (this.right) {
                this.right.add(data);
            } else {
                this.right = new Node(data);
            }
        }
    }
};

Node.prototype.contains = function(data) {
    if (!data) {
        return false;
    }

    if (this.data !== data) {
        if (data < this.data) {
            return this.left ? this.left.contains(data) : false;
        } else {
            return this.right ? this.right.contains(data) : false;
        }
    } else {
        return true;
    }
};

Node.prototype.validate = function() {
    return validationHelper(this, null, null);
}

function validationHelper(node, minValue, maxValue) {

    // If max is specified, current node value cannot be greater than max
    if (maxValue && node.data > maxValue) {
        return false;
    }

    // If min is specified, current node value cannot be less than min
    if (minValue && node.data < minValue) {
        return false;
    }

    // Processing left so pass a maximum valid value
    if (node.left && !validationHelper(node.left, null, node.data)) {
        return false;
    }

    // Processing right so pass a minimum valid value
    if (node.right && !validationHelper(node.right, node.data, null)) {
        return false;
    }

    return true;
}

function testBST() {
    var bst = new BST(27);
    var root = bst.root;

    root.add(56);
    root.add(58);
    root.add(30);
    root.add(57);

    print("BST: " + JSON.stringify(bst));
    print("root.contains(30) = " + root.contains(30));
    print("root.contains(-12) = " + root.contains(-12));
    print("root.contains(27) = " + root.contains(27));
    print("root.contains(55) = " + root.contains(55));

    print("BST is valid = " + root.validate());
}

File: src/main/java/com/devxpress/BST.java
package com.devxpress;

public class BST<T extends Comparable<T>> {

    private final Node<T> root;

    public BST(T data) {
        this.root = new Node<>(data);
    }

    public Node<T> getRoot() {
        return this.root;
    }

    @Override
    public String toString() {
        return "BST{" +
                "root=" + root +
                '}';
    }

    public static class Node<T extends Comparable<T>> {
        private final T data;
        private Node<T> left = null;
        private Node<T> right = null;

        public Node(T data) {
            this.data = data;
        }

        public void add(T data) {
            if (data != null && data.compareTo(this.data) != 0) {
                if (data.compareTo(this.data) < 0) {
                    if (this.left != null) {
                        this.left.add(data);
                    } else {
                        this.left = new Node<>(data);
                    }
                } else {
                    if (this.right != null) {
                        this.right.add(data);
                    } else {
                        this.right = new Node<>(data);
                    }
                }
            }
        }

        public boolean contains(T data) {
            if (data == null) {
                return false;
            }

            if (data.compareTo(this.data) != 0) {
                if (data.compareTo(this.data) < 0) {
                    return this.left != null && this.left.contains(data);
                } else {
                    return this.right != null && this.right.contains(data);
                }
            } else {
                return true;
            }
        }

        public boolean validate() {
            return validationHelper(this, null, null);
        }

        private boolean validationHelper(Node<T> node, T minValue, T maxValue) {

            // If max is specified, current node value cannot be greater than max
            if (maxValue != null && node.data.compareTo(maxValue) > 0) {
                return false;
            }

            // If min is specified, current node value cannot be less than min
            if (minValue != null && node.data.compareTo(minValue) < 0) {
                return false;
            }

            // Processing left so pass a maximum valid value
            if (node.left != null && !validationHelper(node.left, null, node.data)) {
                return false;
            }

            // Processing right so pass a minimum valid value
            if (node.right != null && !validationHelper(node.right, node.data, null)) {
                return false;
            }

            return true;
        }

        @Override
        public String toString() {
            return "Node{" +
                    "data=" + data +
                    ", L=" + left +
                    ", R=" + right +
                    '}';
        }
    }
}
